test/runner.spec.js

```diff
@@ -433,4 +433,87 @@ describe('Runner', function () {
       });
     });
   });
+
+  describe('.run', function () {
+    describe('"test end" event', function () {
+      it('should emit "test end" event for each test', function (done) {
+        var finishedTestTitles = [];
+
+        suite.addTest(new Test('first test', noop));
+        suite.addTest(new Test('second (failed) test', function () {
+          expect(1 + 1).to.equal(3);
+        }));
+        suite.addTest(new Test('third test', noop));
+
+        runner.on('test end', function (test) {
+          finishedTestTitles.push(test.title);
+        });
+
+        runner.run(function () {
+          finishedTestTitles.should.deepEqual([
+            'first test',
+            'second (failed) test',
+            'third test'
+          ]);
+          done();
+        });
+      });
+
+      it('should have a proper duration field', function (done) {
+        var finishedTestDuration = [];
+
+        function doSomethingForSeveralMilliseconds () {
+          var result = 0;
+          for (var i = 0; i < 10000000; i++) {
+            result = i;
+          }
+
+          expect(result).to.equal(100001);
+        }
+
+        function doneAfterTimeout (done) {
+          setTimeout(function () {
+            var result = 0;
+            for (var i = 0; i < 10000000; i++) {
+              result = i;
+            }
+            if (result > 0) {
+              done();
+            }
```

Comment:
If the test were supposed to be able to fail, a specific error would be better than a timeout unless a timeout is the specific failure being tested; and in this case it appears the test is not supposed to be able to fail. Either way, don't call `done` conditionally with the alternative being to niether call `done` nor throw.